File: Solution/WebSolution/AssessmentPortal/Monolithic/Monolithic_Controller/AssessmentAPI/Controllers/AssessmentController.cs
using System.Diagnostics;
using Microsoft.AspNetCore.Mvc;
using Newtonsoft.Json;
using Transflower.TFLAssessment.Entities;
using  Transflower.TFLAssessment.Services.Interfaces;
using Transflower.TFLAssessment.Entities.Models;

namespace Transflower.TFLAssessment.Controllers;

//Controller is now responsible to handle HTTP Requests

[ApiController]
[Route("api/[controller]")]
public class AssessmentController : ControllerBase
{
    private readonly IAssessmentService _svc;
    private readonly ILogger<AssessmentController> _logger;
    public AssessmentController(IAssessmentService service, ILogger<AssessmentController> logger)
    {
        // Initialize with some sample data
        _svc = service;
        _logger = logger;

    }
    //http://localhost:5238/api/assessment/1
    [HttpGet("{id}")]
    public async Task<IActionResult> GetDetails(int id)
    {
        Assessment theAssessment = await _svc.GetDetails(id);
        _logger.LogInformation("Get details method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(theAssessment);
    }

    // GET: api/assessment
    [HttpGet("creationdate/fromDate/{fromDate}/toDate/{toDate}")]
    public async Task<IActionResult> GetAll(DateTime fromDate, DateTime toDate)
    {
        List<Assessment> assessments = await _svc.GetAll(fromDate, toDate);
        _logger.LogInformation("Get all Assessments method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(assessments);
    }

    // GET: get all assessments
    //http://localhost:5238/api/assessment/assessments
    [HttpGet("assessments")]
    public async Task<IActionResult> GetAllAssesment()
    {
        List<Assessment> assessments = await _svc.GetAllTests();
        _logger.LogInformation("Get all tests method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(assessments);
    }

    //http://localhost:5238/api/assessment/employees
    [HttpGet("employees")]
    public async Task<IActionResult> GetAllEmployees()
    {
        List<Employee> employees = await _svc.GetAllEmployees();
        _logger.LogInformation("Get all employee method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(employees);
    }

    //http://localhost:5238/api/assessment/employee/{userId}
    [HttpGet("employee/{userId}")]
    public async Task<IActionResult> GetEmployeeById(int userId)
    {
        Employee employee = await _svc.GetEmployeeById(userId);
        _logger.LogInformation("Get all employee method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(employee);
    }

    //http://localhost:5238/api/assessment/subjects
    [HttpGet("subjects")]
    public async Task<IActionResult> GetAllSubjects()
    {
        List<Subject> subjects = await _svc.GetAllSubjects();
        _logger.LogInformation("Get all subject method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(subjects);
    }


    //http://localhost:5238/api/assessment/criterias
    [HttpGet("criterias")]
    public async Task<IActionResult> GetEvalutionCriterias()
    {
        List<EvaluationCriteria> criterias = await _svc.GetEvalutionCriterias();
        _logger.LogInformation("Get evaluation criteria method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(criterias);
    }

    //http://localhost:5238/api/assessment/criterias/subjects/1
    [HttpGet("criterias/subjects/{subjectId}")]
    public async Task<IActionResult> GetEvalutionCriteriasBySubject(int subjectId)
    {
        List<EvaluationCriteria> criterias = await _svc.GetEvalutionCriteriasBySubject(subjectId);
        _logger.LogInformation("Get  evaluation criteria by subject method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(criterias);
    }



    //http://localhost:5238/api/assessment/subjectexperts/2
    [HttpGet("subjectexperts/{smeId}")]
    public async Task<IActionResult> GetAllBySubjectMatterExpert(int smeId)
    {
        List<Assessment> assessments = await _svc.GetAllBySubjectMatterExpert(smeId);
        _logger.LogInformation("Get all subject matter experties method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(assessments);
    }

    //http://localhost:5238/api/Assessment/createtest
    [HttpPost("createtest")]
    public async Task<IActionResult> CreateTest([FromBody] CreateTestRequest request)
    {
        Console.WriteLine("Inside Create Test Controller");
        bool status = await _svc.CreateTest(request);
        _logger.LogInformation("Create test method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        if (status)
        {
            return Ok(new { message = "Test created successfully" });
        }
        else
        {
            return BadRequest(new { message = "Failed to create test" });
        }
    }

    //http://localhost:5238/api/Assessment/addquestion/assessments/1/questions/10
    [HttpPost("addquestion/assessments/{assessmentId}/questions/{questionId}")]
    public async Task<IActionResult> AddQuestion(int assessmentId, int questionId)
    {
        bool status = await _svc.AddQuestion(assessmentId, questionId);
        _logger.LogInformation("Add question method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(status);
    }


    // http://localhost:5238/api/Assessment/addmultiplequestions/assessments/1
    [HttpPost("addmultiplequestions/assessments/{assessmentId}")]
    public async Task<IActionResult> AddQuestions(int assessmentId, List<QuestionBank> questions)
    {
        // Log the incoming data for debugging
        _logger.LogInformation("Received Questions: {questions}", JsonConvert.SerializeObject(questions));

        bool status = await _svc.AddQuestions(assessmentId, questions);
        _logger.LogInformation("Add multiple questions method invoked at {DT}", DateTime.UtcNow.ToLongTimeString());

        return Ok(status);
    }

    //http://localhost:5238/api/Assessment/1/questions/9
    [HttpDelete("{assessmentId}/questions/{questionId}")]
    public async Task<IActionResult> RemoveQuestion(int assessmentId, int questionId)
    {
        bool status = await _svc.RemoveQuestion(assessmentId, questionId);
        _logger.LogInformation("Remove question method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(status);
    }

    //http://localhost:5238/api/Assessment/1/duration/40
    [HttpPut("{assessmentId}/duration/{duration}")]
    public async Task<IActionResult> ChangeDuration(int assessmentId, string duration)
    {
        bool status = await _svc.ChangeDuration(assessmentId, duration);
        _logger.LogInformation("Change duration method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(status);
    }

    //http://localhost:5238/api/Assessment/1/reschedule/2024-01-01
    [HttpPut("{assessmentId}/reschedule/{date}")]
    public async Task<IActionResult> Reschedule(int assessmentId, DateTime date)
    {
        bool status = await _svc.Reschedule(assessmentId, date);
        _logger.LogInformation("Reschedule method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(status);
    }

    //http://localhost:5238/api/Assessment/deletequestions
    [HttpDelete("deletequestions")]
    public async Task<IActionResult> RemoveQuestions(int[] testQuestions)
    {
        bool status = await _svc.RemoveQuestions(testQuestions);
        _logger.LogInformation("Remove questions method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(status);
    }



    // add test and there questions
    //http://localhost:5238/api/Assessment/addtest
    [HttpPost("addtest")]
    public async Task<IActionResult> AddTest([FromBody] CreateTestWithQuestions request)
    {
        var testId = await _svc.CreateTestWithQuestionsAsync(request);
        if (testId == 0)
        {
            _logger.LogError("Failed to create test at {DT}", DateTime.UtcNow.ToLongTimeString());
            return BadRequest(new { message = "Failed to create test" });
        }
        return Ok(new { message = "Test created", testId });
    }

    //http://localhost:5238/api/Assessment/allquestionsbysubject/1
    [HttpGet("allquestionsbysubject/{subjectId}")]
    public async Task<IActionResult> GetAllQuestionsBySubject(int subjectId)
    {
        List<SubjectQuestions> questions = await _svc.GetAllQuestionsBySubject(subjectId);
        if (questions == null || questions.Count == 0)
        {
            _logger.LogWarning("No questions found for subject ID {SubjectId} at {DT}", subjectId, DateTime.UtcNow.ToLongTimeString());
            return NotFound(new { message = "No questions found for the specified subject." });
        }
        _logger.LogInformation("Get all questions by subject method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(questions);
    }

    //http://localhost:5238/api/Assessment/getsme/1
    [HttpGet("getsme/{subjectId}")]
    public async Task<IActionResult> GetSmeBySubject(int subjectId)
    {
        List<Employee> smeList = await _svc.GetSmeBySubject(subjectId);
        if (smeList == null || smeList.Count == 0)
        {
            _logger.LogWarning("No SME found for subject ID {SubjectId} at {DT}", subjectId, DateTime.UtcNow.ToLongTimeString());
            return NotFound(new { message = "No SME found for the specified subject." });
        }
        _logger.LogInformation("Get SME by subject method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(smeList);
    }

    //http://localhost:5238/api/Assessment/getalltest/from/2024-01-01/to/2024-12-31
    [HttpGet("getalltest/from/{fromDate}/to/{toDate}")]
    public async Task<IActionResult> GetAllTests(DateTime fromDate, DateTime toDate)
    {
        List<Test> assessments = await _svc.GetAllTests(fromDate, toDate);
        if (assessments == null || assessments.Count == 0)
        {
            _logger.LogWarning("No tests found between {FromDate} and {ToDate} at {DT}", fromDate, toDate, DateTime.UtcNow.ToLongTimeString());
            return NotFound(new { message = "No tests found for the specified date range." });
        }
        _logger.LogInformation("Get all tests method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(assessments);
    }

    // Get all test details by test ID
    //http://localhost:5238/api/Assessment/testdetails/{testId}
    [HttpGet("testdetails/{testId}")]
    public async Task<IActionResult> GetTestDetails(int testId)
    {
        var testDetails = await _svc.GetTestDetails(testId);
        if (testDetails == null)
        {
            _logger.LogWarning("No test details found for test ID {TestId} at {DT}", testId, DateTime.UtcNow.ToLongTimeString());
            return NotFound(new { message = "No test details found for the specified test ID." });
        }
        _logger.LogInformation("Get test details method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(testDetails);
    }
    //evaluation criteria id give me all question base on evaluation criteria
    //http://localhost:5238/api/Assessment/questionsbycriteria/{EvaluationCriteriaId}
    [HttpGet("questionsbycriteria/{EvaluationCriteriaId}")]
    public async Task<IActionResult> GetQuestionsByEvaluationCriteriaId(int EvaluationCriteriaId)
    {
        var questionDetails = await _svc.GetQuestionsByEvaluationCriteriaId(EvaluationCriteriaId);
        if (questionDetails == null)
        {
            _logger.LogWarning("No test details found for test ID {TestId} at {DT}", EvaluationCriteriaId, DateTime.UtcNow.ToLongTimeString());
            return NotFound(new { message = "No test details found for the specified test ID." });
        }
        _logger.LogInformation("Get test details method invoked at  {DT}", DateTime.UtcNow.ToLongTimeString());
        return Ok(questionDetails);
    }
    // update question
    //http://localhost:5238/api/Assessment/updatequestion/{questionId}
    [HttpPut("updatequestion/{questionId}")]
    public async Task<IActionResult> UpdateQuestion(int questionId, [FromBody] Question question)
    {
        if (question == null || question.Id != questionId)
        {
            _logger.LogError("Invalid question data provided at {DT}", DateTime.UtcNow.ToLongTimeString());
            return BadRequest(new { message = "Invalid question data provided." });
        }

        // Assuming the service has an UpdateQuestion method
        bool status = await _svc.UpdateQuestion(question);
        if (!status)
        {
            _logger.LogError("Failed to update question with ID {QuestionId} at {DT}", questionId, DateTime.UtcNow.ToLongTimeString());
            return BadRequest(new { message = "Failed to update question." });
        }

        _logger.LogInformation("Question with ID {QuestionId} updated successfully at {DT}", questionId, DateTime.UtcNow.ToLongTimeString());
        return Ok(new { message = "Question updated successfully." });
    }
}

File: Solution/WebSolution/Java/AssessmentAPI/src/main/java/com/transflower/tflassessment/controllers/AssessmentController.java
package com.transflower.tflassessment.controllers;

import java.time.LocalDateTime;
import java.util.Date;
import java.util.List;
import java.util.concurrent.CompletableFuture;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import com.transflower.tflassessment.entities.Assessment;
import com.transflower.tflassessment.entities.CreateTestRequest;
import com.transflower.tflassessment.entities.CreateTestWithQuestions;
import com.transflower.tflassessment.entities.Employee;
import com.transflower.tflassessment.entities.EvaluationCriteria;
import com.transflower.tflassessment.entities.Question;
import com.transflower.tflassessment.entities.QuestionBank;
import com.transflower.tflassessment.entities.Subject;
import com.transflower.tflassessment.entities.SubjectQuestion;
import com.transflower.tflassessment.entities.Test;
import com.transflower.tflassessment.entities.TestEmployeeDetails;
import com.transflower.tflassessment.entities.TestEmployeeRequest;
import com.transflower.tflassessment.entities.TestStatusUpdate;
import com.transflower.tflassessment.entities.TestWithQuestions;
import com.transflower.tflassessment.services.AssessmentService;

@RestController
public class AssessmentController {

    @Autowired
    private AssessmentService svc;

    @GetMapping("/api/assessment/{id}")
    public CompletableFuture<Assessment> getDetails(@PathVariable("id") int id) {
        return svc.getDetails(id);
    }

    @GetMapping("/api/assessment/creationdate/fromDate/{fromDate}/toDate/{toDate}")
    public CompletableFuture<List<Assessment>> getAll( @PathVariable("fromDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDateTime fromDate,
            @PathVariable("toDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDateTime toDate) {
        return svc.getAll(fromDate, toDate);
    }

    @GetMapping("/api/assessment/assessments")
    public CompletableFuture<List<Assessment>> getAllTests() {
        return svc.getAllTests();
    }

    @GetMapping("/api/assessment/employees")
    public CompletableFuture<List<Employee>> getAllEmployees() {
        return svc.getAllEmployees();
    }

    @GetMapping("/api/assessment/employee/{userId}")
    public CompletableFuture<Employee> getEmployeeById(@PathVariable("userId") int userId) {
        return svc.getEmployeeById(userId);
    }

    @GetMapping("/api/assessment/subjects")
    public CompletableFuture<List<Subject>> getAllSubjects() {
        return svc.getAllSubjects();
    }

    @GetMapping("/api/assessment/criterias")
    public CompletableFuture<List<EvaluationCriteria>> getEvaluationCriterias() {
        return svc.getEvaluationCriterias();
    }

    @GetMapping("/api/assessment/criterias/subjects/{subjectId}")
    public CompletableFuture<List<EvaluationCriteria>> getEvaluationCriteriasBySubject(@PathVariable("subjectId") int subjectId) {
        return svc.getEvaluationCriteriasBySubject(subjectId);
    }

    @GetMapping("/api/assessment/subjectexperts/{smeId}")
    public CompletableFuture<List<Assessment>> getAllBySubjectMatterExpert(@PathVariable("smeId") int smeId) {
        return svc.getAllBySubjectMatterExpert(smeId);
    }

    @PostMapping("/api/assessment/createtest")
    public CompletableFuture<Boolean> createTest(@RequestBody CreateTestRequest request) {
        return svc.createTest(request);
    }

    @PostMapping("/api/assessment/addquestion/assessments/{assessmentId}/questions/{questionId}")
    public CompletableFuture<Boolean> addQuestion(@PathVariable("assessmentId") int assessmentId, @PathVariable("questionId") int questionId) {
        return svc.addQuestion(assessmentId, questionId);
    }

    @PostMapping("/api/assessment/addmultiplequestions/assessments/{assessmentId}")
    public CompletableFuture<Boolean> addQuestions(@PathVariable("assessmentId") int assessmentId, @RequestBody List<QuestionBank> questions) {
        return svc.addQuestions(assessmentId, questions);
    }

    @DeleteMapping("/api/assessment/{assessmentId}/questions/{questionId}")
    public CompletableFuture<Boolean> removeQuestion(@PathVariable("assessmentId") int assessmentId,  @PathVariable("questionId") int questionId) {
        return svc.removeQuestion(assessmentId, questionId);
    }

    @PutMapping("/api/assessment/{assessmentId}/duration/{duration}")
    public CompletableFuture<Boolean> changeDuration(@PathVariable("assessmentId") int assessmentId, @PathVariable("duration") String duration) {
        return svc.changeDuration(assessmentId, duration);
    }

    @PutMapping("/api/assessment/{assessmentId}/reschedule/{date}")
    public CompletableFuture<Boolean> reschedule(@PathVariable("assessmentId") int assessmentId,@PathVariable("date") Date date) {
        return svc.reschedule(assessmentId, date);
    }

    @DeleteMapping("/api/assessment/deletequestions")
    public CompletableFuture<Boolean> removeQuestions(@RequestBody int[] testQuestions) {
        return svc.removeQuestions(testQuestions);
    }

    @PostMapping("/api/assessment/addtest")
    public CompletableFuture<Integer> createTestWithQuestions(@RequestBody CreateTestWithQuestions createTestWithQuestions) {
        return svc.createTestWithQuestions(createTestWithQuestions);
    }

    @GetMapping("/api/assessment/allquestionsbysubject/{subjectId}")
    public CompletableFuture<List<SubjectQuestion>> getAllQuestionsBySubject(@PathVariable("subjectId") int subjectId) {
        return svc.getAllQuestionsBySubject(subjectId);
    }

    @GetMapping("/api/assessment/getsme/{subjectId}")
    public CompletableFuture<List<Employee>> getSmeBySubject(@PathVariable("subjectId") int subjectId) {
        return svc.getSmeBySubject(subjectId);
    }

    @GetMapping("/api/assessment/getalltest/from/{fromDate}/to/{toDate}")
    public CompletableFuture<List<Test>> getAllTests(
            @PathVariable("fromDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) Date fromDate,
               @PathVariable("toDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) Date toDate) {
        return svc.getAllTests(fromDate, toDate);
    }

    @GetMapping("/api/assessment/testdetails/{testId}")
    public CompletableFuture<TestWithQuestions> getTestDetails(@PathVariable("testId") int testId) {
        return svc.getTestDetails(testId);
    }

    @GetMapping("/api/assessment/questionsbycriteria/{evaluationCriteriaId}")
    public CompletableFuture<List<Question>> getQuestionsByEvaluationCriteriaId(
            @PathVariable("evaluationCriteriaId") int evaluationCriteriaId) {
        return svc.getQuestionsByEvaluationCriteriaId(evaluationCriteriaId);
    }

    @PutMapping("/api/assessment/updatequestion/{questionId}")
    public CompletableFuture<Boolean> updateQuestion(@PathVariable int questionId, @RequestBody Question question) {
        question.setId(questionId);
        return svc.updateQuestion(question);
    }

    @PutMapping("/api/assessment/updateteststatus/{testId}")
    public CompletableFuture<Boolean> updateTestStatus(@PathVariable("testId") int testId,  @RequestBody TestStatusUpdate status) {
        return svc.updateTestStatus(testId, status);
    }

    @PostMapping("/api/assessment/addemployees")
    public CompletableFuture<Boolean> addEmployeesToTest(@RequestBody TestEmployeeRequest wrapper) {
        return svc.addEmployeesToTest(wrapper.getRequest(), wrapper.getCandidateTestDetails());
    }

    @GetMapping("/api/assessment/alltestbyempid/{empId}")
    public CompletableFuture<List<TestEmployeeDetails>> getAllTestByEmpId(@PathVariable("empId") int empId) {
        return svc.getAllTestByEmpId(empId);
    }
}
